addons/l10n_in/models/company.py

```diff
@@ -12,3 +11,54 @@ class ResCompany(models.Model):
         ],
         string="HSN Code Digit",
     )
+
+    def write(self, vals):
+        old_state_vals = {c.id: c.state_id.id for c in self}
+        res = super().write(vals)
+        if vals.get('state_id'):
+            # Update Fiscal Positions for companies setting up state for the first time
+            companies_need_update_fp = self.filtered(lambda c:\
+                c.account_fiscal_country_id.code == 'IN' and\
+                not old_state_vals.get(c.id) and\
+                c.state_id and\
+                c.state_id.country_id.code == 'IN')
+            for company in companies_need_update_fp:
+                company._update_l10n_in_fiscal_position()
+        return res
+
+    def _get_l10n_in_fiscal_positions_vals_by_ref(self):
+        self.ensure_one()
+        return {
+            'fiscal_position_in_intra_state': {
+                'sequence': 1,
+                'auto_apply': True,
+                'country_id': self.country_id.id,
+                'state_ids': [Command.set(self.state_id.ids)]
+            },
+            'fiscal_position_in_inter_state': {
+                'sequence': 2,
+                'auto_apply': True,
+            },
+        }
+
+    def _update_l10n_in_fiscal_position(self):
+        self.ensure_one()
+        for _xml_id, vals in self._get_l10n_in_fiscal_positions_vals_by_ref().items():
```

Comment:
Why so complicated?  We need to just update that one fiscal position with the correct state.  

Of course, the fiscal positions can be shared among branches and that complicates matters...  I do not see how that can work here.  (suppose your parent company is state A and your child company is state B, then state A will still be seen as intrastate).  Another way might be to change the company domain just for this fiscal position... (but that is hacky)

Reply:
Well I still need to work on this, it is not going to be an easy job for sure, may be only inheriting `write` method won't be enough

Reply:
@hamo-odoo I think there is a way to do it (to avoid the problem if selecting the wrong state fiscal position from a parent branch) if you set the sequence field as well on the fiscal position (child branch with higher priority than parent branch), but we are still thinking if we could do something in the fiscal position auto_apply logic itself.